### PR TITLE
CR-1130186 Refactor verify test to use XRT native API instead of OpenCL

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
https://jira.xilinx.com/browse/CR-1130186
The verify test failure is not being passed back correctly to the xbutil tool. The test uses the OpenCl framework which prevents immediate feedback and suppresses task return codes which makes handing back valid failure information difficult. As a result the DMA test would fail due to the verify kernel test task not succeeding. 

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Discovered via test case failure.

#### How problem was solved, alternative solutions (if any) and why they were rejected
The problem was solved by migrating the verify executable to use the XRT native API which does not suffer from the same limitation.

#### What has been tested and how, request additional testing if necessary
Ubutntu 20 \ u55c
Detecting failure
```
dbenusov@xsjpranjal01:/proj/rdi/staff/dbenusov$ ./XRT/build/Debug/opt/xilinx/xrt/test/validate.exe -p /opt/xilinx/firmware/u55c/gen3x16-xdma/base/test/ -d 0000:17:00.1
Trying to program device...
Device program successful!
terminate called after throwing an instance of 'xrt_core::system_error'
  what():  unable to sync BO: Input/output error
Aborted (core dumped)
dbenusov@xsjpranjal01:/proj/rdi/staff/dbenusov$ echo $?
134
dbenusov@xsjpranjal01:/proj/rdi/staff/dbenusov$ xbutil validate -d 17:00 -r verify
Starting validation for 1 devices

Validate Device           : [0000:17:00.1]
    Platform              : xilinx_u55c_gen3x16_xdma_base_3
    SC Version            : 7.1.17
    Platform ID           : 97088961-FEAE-DA91-52A2-1D9DFD63CCEF
-------------------------------------------------------------------------------
Test 1 [0000:17:00.1]     : verify
    Error(s)              :
                            terminate called after throwing an instance of
                            'xrt_core::system_error'
                              what():  unable to sync BO: Input/output error
    Test Status           : [FAILED]
-------------------------------------------------------------------------------
Validation failed. Please run the command '--verbose' option for more details
```
Detecting success
```
dbenusov@xsjpranjal01:/proj/rdi/staff/dbenusov/XRT$ ./build/Debug/opt/xilinx/xrt/test/validate.exe -p /opt/xilinx/firmware/u55c/gen3x16-xdma/base/test/ -d 0000:17:00.1
Trying to program device...
Device program successful!
TEST PASSED
dbenusov@xsjpranjal01:/proj/rdi/staff/dbenusov/XRT$ xbutil validate -d 17:00 -r verify
Starting validation for 1 devices

Validate Device           : [0000:17:00.1]
    Platform              : xilinx_u55c_gen3x16_xdma_base_3
    SC Version            : 7.1.17
    Platform ID           : 97088961-FEAE-DA91-52A2-1D9DFD63CCEF
-------------------------------------------------------------------------------
Test 1 [0000:17:00.1]     : verify
    Test Status           : [PASSED]
-------------------------------------------------------------------------------
Validation completed. Please run the command '--verbose' option for more details
```
#### Documentation impact (if any)
None

### CODE REVIEW

File: tests/validate/verify_test/CMakeLists.txt [hunk 1]
```diff
@@ -3,34 +3,24 @@
...
-if (WIN32)
-  set(OpenCL_INCLUDE_DIR ${OCL_ROOT}/include)
-  find_library(OpenCL_LIBRARY
-    NAMES OpenCL
-    HINTS "${OCL_ROOT}/lib")
-  include_directories(${OpenCL_INCLUDE_DIR})
-else()
-  find_package(OpenCL)
-endif(WIN32)
+set(xrt_core_LIBRARY xrt_core)
```
Comment (stsoe) [inline tests/validate/verify_test/CMakeLists.txt:6]: Should not be needed for a true XRT native API application.

File: tests/validate/verify_test/CMakeLists.txt [hunk 2]
```diff
@@ -3,34 +3,24 @@
...
-  ../common/includes/xcl2/xcl2.cpp
   ../common/includes/cmdparser/cmdlineparser.cpp
   ../common/includes/logger/logger.cpp src/host.cpp
   )
 
 target_link_libraries(${TESTNAME} PRIVATE ${xrt_coreutil_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
 
 if (NOT WIN32)
-  target_link_libraries(${TESTNAME} PRIVATE ${uuid_LIBRARY} pthread ${OpenCL_LIBRARY} )
+  target_link_libraries(${TESTNAME} PRIVATE ${uuid_LIBRARY}  pthread ${xrt_coreutil_LIBRARY} ${xrt_core_LIBRARY})
```
Comment (stsoe) [inline tests/validate/verify_test/CMakeLists.txt:23]: You should be able to remove the xrt_core_LIBRARY

File: tests/validate/verify_test/src/host.cpp [hunk 3]
```diff
@@ -49,96 +53,39 @@ int main(int argc, char** argv) {
...
-    cl::Program program(context, {device}, bins, nullptr, &err);
-    if (err != CL_SUCCESS) {
-        std::cout << "Failed to program device with xclbin file!\n";
-    } else {
-        std::cout << "Device program successful!\n";
-        OCL_CHECK(err, krnl_verify = cl::Kernel(program, "verify", &err));
-    }
+    // Allocate the output buffer to hold the kernel ooutput
+    auto bank_grp_arg0 = krnl.group_id(0);
+    auto output_buffer = xrt::bo(device, sizeof(char) * LENGTH, bank_grp_arg0);
```
Comment (stsoe) [inline tests/validate/verify_test/src/host.cpp:68]: combine

File: tests/validate/verify_test/src/host.cpp [hunk 4]
```diff
@@ -49,96 +53,39 @@ int main(int argc, char** argv) {
...
 
-    OCL_CHECK(err, cl::Buffer d_buf(context, CL_MEM_WRITE_ONLY, sizeof(char) * LENGTH, nullptr, &err));
+    // Run the kernel and store its contents within the allocated output buffer
+    auto run = krnl(output_buffer);
+    run.wait();
 
-    OCL_CHECK(err, err = krnl_verify.setArg(0, d_buf));
+    // Prepare local buffer
+    char received_data[LENGTH];
+    std::memset(received_data, 0, LENGTH);
```
Comment (stsoe) [inline tests/validate/verify_test/src/host.cpp:76]:
```suggestion
    char received_data[LENGTH] = {};
```

**Review by stsoe — CHANGES_REQUESTED** [review]
Looks good and much simpler.  Just a few changes.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: tests/validate/verify_test/src/host.cpp [hunk 5]
```diff
@@ -13,7 +13,11 @@
 * License for the specific language governing permissions and limitations
 * under the License.
 */
-#include "xcl2.hpp"
+
+#include "xrt/xrt_device.h"
+#include "xrt/xrt_bo.h"
+#include "xrt/xrt_kernel.h"
```
Comment (stsoe) [inline tests/validate/verify_test/src/host.cpp:19]: sort

File: tests/validate/verify_test/src/host.cpp [hunk 6]
```diff
@@ -13,7 +13,11 @@
...
+
 #include <boost/filesystem.hpp>
 #include <algorithm>
 #include <vector>
```
Comment (stsoe) [inline tests/validate/verify_test/src/host.cpp:23]: None of these are used in your new code.
But `<cstring>` is for std::memcmp, but you should replace that with std::equal and thus leave inclusion of `<algorithm>`

File: tests/validate/verify_test/CMakeLists.txt [hunk 7]
```diff
@@ -3,34 +3,23 @@
...
   ../common/includes/logger
   )
 
 add_executable(${TESTNAME}
-  ../common/includes/xcl2/xcl2.cpp
   ../common/includes/cmdparser/cmdlineparser.cpp
   ../common/includes/logger/logger.cpp src/host.cpp
   )
 
 target_link_libraries(${TESTNAME} PRIVATE ${xrt_coreutil_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
```
Comment (stsoe) [inline tests/validate/verify_test/CMakeLists.txt:19]: your test no longer uses Boost, not sure if cmdlineparser or logger do?

File: tests/validate/verify_test/src/host.cpp (hunk 6 above)
Comment (dbenusov-xilinx) [reply tests/validate/verify_test/src/host.cpp:23]: I needed to add string.h for strcmp for the argument processing

File: tests/validate/verify_test/src/host.cpp [hunk 8]
```diff
@@ -13,10 +13,13 @@
...
 * under the License.
 */
-#include "xcl2.hpp"
-#include <boost/filesystem.hpp>
+
+#include "xrt/xrt_bo.h"
+#include "xrt/xrt_device.h"
+#include "xrt/xrt_kernel.h"
+
+#include <string.h>
```
[the remark this answers is not in the document]
Comment (dbenusov-xilinx) [reply tests/validate/verify_test/src/host.cpp]: Got it. I had to revert the std::equals change. It failed on the emulation pipeline, I imagine it is something to do with the difference in formatting but memcmp works just as well.

**Comment by salindac** [comment]
Retest this please

**Review by stsoe — APPROVED** [review]

**Pull request merged**